internal/crypto/src/raw_signature/validator.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright 2024 Adobe. All rights reserved.
+// This file is licensed to you under the Apache License,
+// Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+// or the MIT license (http://opensource.org/licenses/MIT),
+// at your option.
+
+// Unless required by applicable law or agreed to in writing,
+// this software is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR REPRESENTATIONS OF ANY KIND, either express or
+// implied. See the LICENSE-MIT and LICENSE-APACHE files for the
+// specific language governing permissions and limitations under
+// each license.
+
+use bcder::Oid;
+use thiserror::Error;
+
+use super::oids::*;
+use crate::SigningAlg;
+
+/// A `RawSignatureValidator` implementation checks a signature encoded using a
+/// specific signature algorithm and a private/public key pair.
+///
+/// IMPORTANT: This signature is typically embedded in a wrapper provided by
+/// another signature mechanism. In the C2PA ecosystem, this wrapper is
+/// typically COSE, but `RawSignatureValidator` does not implement COSE.
+pub trait RawSignatureValidator {
+    /// Return `true` if the signature `sig` is valid for the raw content `data`
+    /// and the public key `public_key`.
+    fn validate(
+        &self,
+        sig: &[u8],
+        data: &[u8],
+        public_key: &[u8],
+    ) -> Result<(), RawSignatureValidationError>;
+}
+
+/// Return a built-in signature validator for the requested signature
+/// algorithm.
+///
+/// Which validators are available may vary depending on the platform and
+/// which crate features were enabled.
+pub fn validator_for_signing_alg(alg: SigningAlg) -> Option<Box<dyn RawSignatureValidator>> {
```

Comment:
I have to think about this.  It may be too restrictive.  There are cases where we validate patterns that don't match our predefined signature types.  For example You can have an EC signature that uses an arbitrary hash length.  They are not limited to for example ES256 where the sig is on curve prime256v1 but the hash is Sha256.  You can use that curve with other hash values.  We see this when validating certs.

Reply:
I see.  We will have to do some testing against those odd cert cases.

Reply:
@mauricefisher64 would love to have new test cases for anything else you're familiar with.